### PR TITLE
Adds 'activity_count_by_workflow' to /project_preferences

### LINKED ISSUE #1796: expose user classification info by workflow
per Notes from Nature conversion request, to show user "badges" (earned by # of classifications) by "expedition" (which are workflows within Panoptes).

per conversation with @camallen:
-will be modifying [this code](https://github.com/zooniverse/Panoptes/blob/master/app/serializers/user_project_preference_serializer.rb#L19)
-maybe as an optional attribute or detecting the presence of a param

### PR DESCRIPTION
**Pull request by zwolf**
Activity counts via UserSeenSubjects are included per workflow to calls made to /project_preferences. There could be the potential for performance issues related to these queries, but the association is already indexed and the non-legacy counts are being generated on the fly already, so this isn't much heavier.

Will close #1796.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

**Comment by camallen** [comment]
specs currently failing

**Comment by zwolf** [comment]
Passing. ID strings vs. ints strikes again, not sure why I didn't catch it.

File: app/models/user_seen_subject.rb [hunk 1]
```diff
@@ -23,6 +22,15 @@ def self.count_user_activity(user_id, workflow_ids=[])
     scope.sum("cardinality(subject_ids)")
   end
 
+  def self.activity_by_workflow(user_id)
+    scope = self.where(user_id: user_id)
```
Comment (camallen) [inline app/models/user_seen_subject.rb]: we should be able to push this to the DB with something like `scope.group(:workflow_id).sum("cardinality(subject_ids)").as_json`

**Comment by camallen** [comment]
Looks good - i'm a bit worried about power users that work on lots of projects and looping over each one, push that to the DB and i think we're good.

**Comment by zwolf** [comment]
Your suggestion worked, so if you're less concerned about stressing the db then that should help. I amended the commit, though in future I'll probably keep committing so you can more easily see the changes, then squash them when I merge. Cool?

**Comment by camallen** [comment]
The DB is optimized for stuff like this, ruby land code not so much. Feel free to commit the way you like (I more often than not amend). We may ask you to squash them if it looks a bit noisy.

File: app/models/user_seen_subject.rb [hunk 2]
```diff
@@ -23,6 +22,11 @@ def self.count_user_activity(user_id, workflow_ids=[])
     scope.sum("cardinality(subject_ids)")
```
Comment (camallen) [inline app/models/user_seen_subject.rb]: could we just summate the per workflow call here now instead of running 2 different db calls?

Comment (camallen) [inline app/models/user_seen_subject.rb]: and cache the one we do call.

Comment (camallen) [inline app/models/user_seen_subject.rb]: Ignore these - too early :(

File: app/serializers/user_project_preference_serializer.rb [hunk 3]
```diff
@@ -16,6 +16,10 @@ def activity_count
     end
   end
 
+  def activity_count_by_workflow
+    UserSeenSubject.activity_by_workflow(@model.user_id)
```
Comment (camallen) [inline app/serializers/user_project_preference_serializer.rb]: So - this is going to return all the workflows for all projects that a user has contributed too, not the ones for the particular project preference, see `user_project_activity` below. 

I'm also wondering if we can't cache the per workflow method here and use it in the below `user_project_activity` via a local cache? They seem to be doing very similar ops.

File: app/serializers/user_project_preference_serializer.rb [hunk 4]
```diff
@@ -16,8 +16,15 @@ def activity_count
...
+    UserSeenSubject.activity_by_workflow(@model.user_id, project_workflows_ids)
+  end
+
   def user_project_activity
-    project_workflows_ids = Workflow.where(project_id: @model.project_id).pluck(:id)
     UserSeenSubject.count_user_activity(@model.user_id, project_workflows_ids)
   end
+
+  def project_workflows_ids
+    Workflow.where(project_id: @model.project_id).pluck(:id)
```
Comment (camallen) [inline app/serializers/user_project_preference_serializer.rb]: should we cache this call? the AR call will probably be cached but we are using it possibly twice per serialized UPP.

**Comment by camallen** [comment]
LGTM - minor comment on the method call cache, up to you on this one.

**Pull request merged**